File: python/euler/bin/p0014.py
# project euler: problem 14


def compute(limit: int) -> str:
    cache = [0] * limit
    cache[1] = 1

    for cur in range(limit // 2, limit):
        if cache[cur] != 0:
            continue

        path = []
        while cur >= limit or cache[cur] == 0:
            path.append(cur)
            if cur % 2 == 0:
                cur //= 2
            else:
                cur = 3 * cur + 1

        steps = cache[cur] + 1
        for i in reversed(path):
            if i < limit:
                cache[i] = steps
            steps += 1

    return str(cache.index(max(cache)))


def solve() -> str:
    return compute(1_000_000)

File: typescript/euler/bin/p0014.ts
// project euler: problem 14

import { max } from "../lib/math.ts";
import { range } from "../lib/util.ts";

export const compute = (limit: number): string => {
  const cache = new Array(limit);
  cache.fill(0);
  cache[1] = 1;

  for (let cur of range(Math.trunc(limit / 2), limit)) {
    if (cache[cur] !== 0) {
      continue;
    }

    const path: number[] = [];
    while (cur >= limit || cache[cur] === 0) {
      path.push(cur);
      if (cur % 2 === 0) {
        cur = Math.trunc(cur / 2);
      } else {
        cur = 3 * cur + 1;
      }
    }

    let steps = cache[cur] + 1;
    path.reverse();
    for (const i of path) {
      if (i < limit) {
        cache[i] = steps;
      }
      steps += 1;
    }
  }

  return String(cache.indexOf(max(cache)));
};

export const solve = (): string => compute(1_000_000);
